File: backend/app/services/bank_transfer_proof_service.py
from datetime import datetime
from decimal import Decimal

from app.constants import BANK_TRANSFER_PROOF_OPEN_STATUSES, PAYMENT_ELIGIBLE_STATUS
from app.extensions import db
from app.models.bank_transfer_proof import BankTransferProof, BankTransferProofPackage
from app.models.package import Package
from app.models.user import User
from app.services.delivery_request_service import compute_payment_total_with_delivery
from app.services.image_upload_service import is_valid_transfer_proof_reference


def _decimal(value) -> Decimal | None:
    if value is None:
        return None
    return Decimal(str(value)).quantize(Decimal("0.01"))


def list_customer_proofs(customer: User, limit: int = 50) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter_by(customer_id=customer.id)
        .order_by(BankTransferProof.submitted_at.desc())
        .limit(limit)
        .all()
    )


def list_open_customer_proofs(customer: User, limit: int = 50) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter(
            BankTransferProof.customer_id == customer.id,
            BankTransferProof.status.in_(BANK_TRANSFER_PROOF_OPEN_STATUSES),
        )
        .order_by(BankTransferProof.submitted_at.desc())
        .limit(limit)
        .all()
    )


def list_pending_customer_proofs(customer: User, limit: int = 50) -> list[BankTransferProof]:
    return list_open_customer_proofs(customer, limit=limit)


def list_open_transfer_proofs(limit: int = 200) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter(BankTransferProof.status.in_(BANK_TRANSFER_PROOF_OPEN_STATUSES))
        .order_by(BankTransferProof.submitted_at.asc())
        .limit(limit)
        .all()
    )


def list_all_transfer_proofs(limit: int = 200) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.order_by(BankTransferProof.submitted_at.desc())
        .limit(limit)
        .all()
    )


def list_transfer_proofs_by_status(status: str, limit: int = 200) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter_by(status=status)
        .order_by(BankTransferProof.submitted_at.desc())
        .limit(limit)
        .all()
    )


def list_pending_transfer_proofs(limit: int = 100) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter_by(status="pending")
        .order_by(BankTransferProof.submitted_at.asc())
        .limit(limit)
        .all()
    )


def list_transfer_proof_history(limit: int = 200) -> list[BankTransferProof]:
    return (
        BankTransferProof.query.filter(~BankTransferProof.status.in_(BANK_TRANSFER_PROOF_OPEN_STATUSES))
        .order_by(BankTransferProof.submitted_at.desc())
        .limit(limit)
        .all()
    )


def count_open_transfer_proofs() -> int:
    return BankTransferProof.query.filter(BankTransferProof.status.in_(BANK_TRANSFER_PROOF_OPEN_STATUSES)).count()


def count_pending_transfer_proofs() -> int:
    return BankTransferProof.query.filter_by(status="pending").count()


def get_transfer_proof(proof_id) -> BankTransferProof | None:
    import uuid

    try:
        pid = uuid.UUID(str(proof_id))
    except ValueError:
        return None
    return BankTransferProof.query.get(pid)


def proof_to_staff_dict(proof: BankTransferProof) -> dict:
    data = proof.to_dict(include_packages=True)
    if proof.customer:
        data["customer_name"] = proof.customer.full_name
        data["shipping_id"] = proof.customer.shipping_id
    return data


def mark_transfer_proof_in_progress(proof: BankTransferProof, staff_user: User) -> BankTransferProof:
    if proof.status != "pending":
        raise ValueError("Only pending transfer proofs can be marked in progress")

    proof.status = "in_progress"
    proof.reviewed_at = datetime.utcnow()
    proof.reviewed_by_id = staff_user.id
    db.session.commit()
    return proof


def confirm_transfer_proof(proof: BankTransferProof, staff_user: User) -> BankTransferProof:
    if proof.status not in BANK_TRANSFER_PROOF_OPEN_STATUSES:
        raise ValueError("Only open transfer proofs can be confirmed")

    package_ids = [str(link.package_id) for link in proof.package_links if link.package_id]
    if package_ids:
        from app.services.payment_service import record_payment_checkout

        record_payment_checkout(
            proof.customer,
            package_ids,
            method="bank_transfer",
            recorded_by=staff_user,
            reference=proof.transfer_reference,
            notes=proof.notes,
        )

    proof.status = "confirmed"
    proof.reviewed_at = datetime.utcnow()
    proof.reviewed_by_id = staff_user.id
    db.session.commit()
    return proof


def reject_transfer_proof(proof: BankTransferProof, staff_user: User) -> BankTransferProof:
    if proof.status not in BANK_TRANSFER_PROOF_OPEN_STATUSES:
        raise ValueError("Only open transfer proofs can be rejected")

    proof.status = "rejected"
    proof.reviewed_at = datetime.utcnow()
    proof.reviewed_by_id = staff_user.id
    db.session.commit()
    return proof


def _validate_proof_packages(customer: User, package_ids: list) -> list[Package]:
    if not package_ids:
        return []

    if len(package_ids) > 50:
        raise ValueError("Cannot link more than 50 packages to one transfer proof")

    packages: list[Package] = []
    seen: set[str] = set()

    for raw_id in package_ids:
        pid = str(raw_id)
        if pid in seen:
            continue
        seen.add(pid)

        package = Package.query.filter_by(id=raw_id, customer_id=customer.id).first()
        if not package:
            raise ValueError("One or more packages were not found on your account")
        if package.billing_status == "paid":
            raise ValueError(f"{package.tracking_number} is already paid")
        if package.status != PAYMENT_ELIGIBLE_STATUS or package.billing_status != "ready":
            raise ValueError(
                f"{package.tracking_number} is not ready for payment yet"
            )
        packages.append(package)

    return packages


def submit_bank_transfer_proof(
    customer: User,
    *,
    proof_object_key: str,
    package_ids: list | None = None,
    transfer_reference: str | None = None,
    amount_jmd=None,
    notes: str | None = None,
) -> BankTransferProof:
    proof_key = (proof_object_key or "").strip()
    if not proof_key:
        raise ValueError("proof_object_key is required")
    if not is_valid_transfer_proof_reference(proof_key):
        raise ValueError("Invalid transfer proof file reference")

    packages = _validate_proof_packages(customer, package_ids or [])

    amount = _decimal(amount_jmd)
    if amount is not None and amount <= 0:
        raise ValueError("amount_jmd must be greater than zero")

    if packages:
        expected = compute_payment_total_with_delivery(customer, [str(p.id) for p in packages])
        expected_total = Decimal(str(expected["total_jmd"]))
        if amount is None:
            amount = expected_total
        elif amount != expected_total:
            raise ValueError(
                f"amount_jmd must match the total due ({float(expected_total):.2f} JMD including delivery fee if applicable)"
            )

    reference = (transfer_reference or "").strip() or None
    note_text = (notes or "").strip() or None
    if note_text and len(note_text) > 500:
        raise ValueError("notes must be 500 characters or fewer")

    proof = BankTransferProof(
        customer_id=customer.id,
        proof_object_key=proof_key,
        transfer_reference=reference,
        amount_jmd=amount,
        notes=note_text,
        status="pending",
        submitted_at=datetime.utcnow(),
    )
    db.session.add(proof)
    db.session.flush()

    for package in packages:
        db.session.add(
            BankTransferProofPackage(
                proof_id=proof.id,
                package_id=package.id,
            )
        )

    db.session.commit()
    return proof

File: server/src/services/bankTransferProofService.ts
import { and, asc, count, desc, eq, inArray, notInArray, sql } from 'drizzle-orm'
import {
  BANK_TRANSFER_PROOF_OPEN_STATUSES,
  DELIVERY_FEE_JMD,
  PAYMENT_ELIGIBLE_STATUS,
  SENDER_BANKS,
} from '../constants.js'
import { db } from '../db/index.js'
import {
  bankTransferProofPackages,
  bankTransferProofs,
  packages,
  users,
  type UserRow,
} from '../db/schema/index.js'
import { bankTransferProofToDict } from '../lib/serializers/bankTransferProof.js'
import { userFullName } from '../lib/serializers/user.js'
import { computePaymentTotalWithDelivery } from './deliveryRequestService.js'
import { isValidTransferProofReference } from './imageUploadService.js'
import { recordPaymentCheckout } from './paymentService.js'

type ProofRow = typeof bankTransferProofs.$inferSelect

async function loadProofDetails(proofId: string) {
  const [proof] = await db.select().from(bankTransferProofs).where(eq(bankTransferProofs.id, proofId)).limit(1)
  if (!proof) return null

  const links = await db
    .select()
    .from(bankTransferProofPackages)
    .where(eq(bankTransferProofPackages.proofId, proof.id))

  const packageLinks = []
  for (const link of links) {
    const [pkg] = await db.select().from(packages).where(eq(packages.id, link.packageId)).limit(1)
    packageLinks.push({ link, pkg: pkg ?? null })
  }

  let reviewedBy = null
  if (proof.reviewedById) {
    ;[reviewedBy] = await db.select().from(users).where(eq(users.id, proof.reviewedById)).limit(1)
  }

  return { proof, packageLinks, reviewedBy: reviewedBy ?? null }
}

export async function countOpenTransferProofs(): Promise<number> {
  const rows = await db
    .select({ value: count() })
    .from(bankTransferProofs)
    .where(inArray(bankTransferProofs.status, [...BANK_TRANSFER_PROOF_OPEN_STATUSES]))
  return Number(rows[0]?.value ?? 0)
}

export async function listCustomerProofs(customer: UserRow, limit = 50) {
  const rows = await db
    .select()
    .from(bankTransferProofs)
    .where(eq(bankTransferProofs.customerId, customer.id))
    .orderBy(desc(bankTransferProofs.submittedAt))
    .limit(limit)

  const result = []
  for (const proof of rows) {
    const details = await loadProofDetails(proof.id)
    if (!details) continue
    result.push(
      bankTransferProofToDict(details.proof, {
        includePackages: true,
        packageLinks: details.packageLinks,
        reviewedBy: details.reviewedBy,
      }),
    )
  }
  return result
}

async function validateProofPackages(customer: UserRow, packageIds: unknown[]) {
  if (!packageIds.length) return []

  if (packageIds.length > 50) throw new Error('Cannot link more than 50 packages to one transfer proof')

  const pkgs: Array<typeof packages.$inferSelect> = []
  const seen = new Set<string>()

  for (const rawId of packageIds) {
    const pid = String(rawId)
    if (seen.has(pid)) continue
    seen.add(pid)

    const [pkg] = await db
      .select()
      .from(packages)
      .where(eq(packages.id, pid))
      .limit(1)
    if (!pkg || pkg.customerId !== customer.id) {
      throw new Error('One or more packages were not found on your account')
    }
    if (pkg.billingStatus === 'paid') throw new Error(`${pkg.trackingNumber} is already paid`)
    if (pkg.status !== PAYMENT_ELIGIBLE_STATUS || pkg.billingStatus !== 'ready') {
      throw new Error(`${pkg.trackingNumber} is not ready for payment yet`)
    }
    pkgs.push(pkg)
  }
  return pkgs
}

async function expectedProofTotal(
  customer: UserRow,
  pkgs: Array<typeof packages.$inferSelect>,
  includeDeliveryFee: boolean,
): Promise<number> {
  const expected = await computePaymentTotalWithDelivery(
    customer,
    pkgs.map((p) => p.id),
  )
  let total = expected.total_jmd
  if (includeDeliveryFee && expected.delivery_fee_jmd === 0) {
    total += parseFloat(DELIVERY_FEE_JMD)
  }
  return Math.round(total * 100) / 100
}

export async function submitBankTransferProof(
  customer: UserRow,
  opts: {
    proofObjectKey: string
    packageIds?: unknown[]
    transferReference?: string | null
    senderBank?: string | null
    amountJmd?: unknown
    includeDeliveryFee?: boolean
    notes?: string | null
  },
) {
  const proofKey = (opts.proofObjectKey ?? '').trim()
  if (!proofKey) throw new Error('proof_object_key is required')
  if (!isValidTransferProofReference(proofKey)) throw new Error('Invalid transfer proof file reference')

  const pkgs = await validateProofPackages(customer, opts.packageIds ?? [])

  let amount: number | null =
    opts.amountJmd != null ? Math.round(parseFloat(String(opts.amountJmd)) * 100) / 100 : null
  if (amount != null && amount <= 0) throw new Error('amount_jmd must be greater than zero')

  if (pkgs.length) {
    const expectedTotal = await expectedProofTotal(customer, pkgs, Boolean(opts.includeDeliveryFee))
    if (amount == null) amount = expectedTotal
    else if (amount !== expectedTotal) {
      throw new Error(
        `amount_jmd must match the total due (${expectedTotal.toFixed(2)} JMD including delivery fee if applicable)`,
      )
    }
  }

  const reference = (opts.transferReference ?? '').trim() || null
  const bank = (opts.senderBank ?? '').trim().toLowerCase() || null
  if (!bank) throw new Error('sender_bank is required')
  if (!(SENDER_BANKS as readonly string[]).includes(bank)) throw new Error('Invalid sender bank')

  const noteText = (opts.notes ?? '').trim() || null
  if (noteText && noteText.length > 500) throw new Error('notes must be 500 characters or fewer')

  const [proof] = await db
    .insert(bankTransferProofs)
    .values({
      customerId: customer.id,
      proofObjectKey: proofKey,
      transferReference: reference,
      senderBank: bank,
      amountJmd: amount != null ? amount.toFixed(2) : null,
      includeDeliveryFee: Boolean(opts.includeDeliveryFee),
      notes: noteText,
      status: 'pending',
      submittedAt: new Date(),
    })
    .returning()

  for (const pkg of pkgs) {
    await db.insert(bankTransferProofPackages).values({
      proofId: proof!.id,
      packageId: pkg.id,
    })
  }

  const details = await loadProofDetails(proof!.id)
  return bankTransferProofToDict(details!.proof, {
    includePackages: true,
    packageLinks: details!.packageLinks,
    reviewedBy: details!.reviewedBy,
  })
}

export async function listPendingCustomerProofs(customer: UserRow, limit = 50) {
  return listOpenCustomerProofs(customer, limit)
}

export async function listOpenCustomerProofs(customer: UserRow, limit = 50) {
  const openStatuses = [...BANK_TRANSFER_PROOF_OPEN_STATUSES]
  return db
    .select()
    .from(bankTransferProofs)
    .where(and(eq(bankTransferProofs.customerId, customer.id), inArray(bankTransferProofs.status, openStatuses)))
    .orderBy(desc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function listOpenTransferProofs(limit = 100) {
  const openStatuses = [...BANK_TRANSFER_PROOF_OPEN_STATUSES]
  return db
    .select()
    .from(bankTransferProofs)
    .where(inArray(bankTransferProofs.status, openStatuses))
    .orderBy(asc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function listAllTransferProofs(limit = 100) {
  return db
    .select()
    .from(bankTransferProofs)
    .orderBy(desc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function listTransferProofsByStatus(status: string, limit = 100) {
  return db
    .select()
    .from(bankTransferProofs)
    .where(eq(bankTransferProofs.status, status))
    .orderBy(desc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function listPendingTransferProofs(limit = 100) {
  return db
    .select()
    .from(bankTransferProofs)
    .where(eq(bankTransferProofs.status, 'pending'))
    .orderBy(asc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function listTransferProofHistory(limit = 100) {
  const openStatuses = [...BANK_TRANSFER_PROOF_OPEN_STATUSES]
  return db
    .select()
    .from(bankTransferProofs)
    .where(notInArray(bankTransferProofs.status, openStatuses))
    .orderBy(desc(bankTransferProofs.submittedAt))
    .limit(limit)
}

export async function getTransferProof(proofId: string): Promise<ProofRow | null> {
  try {
    const [proof] = await db.select().from(bankTransferProofs).where(eq(bankTransferProofs.id, proofId)).limit(1)
    return proof ?? null
  } catch {
    return null
  }
}

export async function proofToStaffDict(proof: ProofRow): Promise<Record<string, unknown>> {
  const details = await loadProofDetails(proof.id)
  if (!details) return bankTransferProofToDict(proof)

  const data = bankTransferProofToDict(details.proof, {
    includePackages: true,
    packageLinks: details.packageLinks,
    reviewedBy: details.reviewedBy,
  })

  const [customer] = await db.select().from(users).where(eq(users.id, proof.customerId)).limit(1)
  if (customer) {
    data.customer_name = userFullName(customer)
    data.shipping_id = customer.shippingId
  }
  return data
}

export async function markTransferProofInProgress(proof: ProofRow, staffUser: UserRow): Promise<ProofRow> {
  if (proof.status !== 'pending') {
    throw new Error('Only pending transfer proofs can be marked in progress')
  }

  const [updated] = await db
    .update(bankTransferProofs)
    .set({
      status: 'in_progress',
      reviewedAt: new Date(),
      reviewedById: staffUser.id,
    })
    .where(eq(bankTransferProofs.id, proof.id))
    .returning()
  return updated!
}

export async function confirmTransferProof(proof: ProofRow, staffUser: UserRow): Promise<ProofRow> {
  if (!(BANK_TRANSFER_PROOF_OPEN_STATUSES as readonly string[]).includes(proof.status)) {
    throw new Error('Only open transfer proofs can be confirmed')
  }

  const links = await db
    .select()
    .from(bankTransferProofPackages)
    .where(eq(bankTransferProofPackages.proofId, proof.id))

  const packageIds = links.map((link) => link.packageId).filter(Boolean)
  if (packageIds.length) {
    const [customer] = await db.select().from(users).where(eq(users.id, proof.customerId)).limit(1)
    if (!customer) throw new Error('Transfer proof customer not found')
    await recordPaymentCheckout(customer, packageIds, {
      method: 'bank_transfer',
      recordedBy: staffUser,
      reference: proof.transferReference,
      notes: proof.notes,
    })
  }

  const [updated] = await db
    .update(bankTransferProofs)
    .set({
      status: 'confirmed',
      reviewedAt: new Date(),
      reviewedById: staffUser.id,
    })
    .where(eq(bankTransferProofs.id, proof.id))
    .returning()
  return updated!
}

export async function rejectTransferProof(proof: ProofRow, staffUser: UserRow): Promise<ProofRow> {
  if (!(BANK_TRANSFER_PROOF_OPEN_STATUSES as readonly string[]).includes(proof.status)) {
    throw new Error('Only open transfer proofs can be rejected')
  }

  const [updated] = await db
    .update(bankTransferProofs)
    .set({
      status: 'rejected',
      reviewedAt: new Date(),
      reviewedById: staffUser.id,
    })
    .where(eq(bankTransferProofs.id, proof.id))
    .returning()
  return updated!
}
